protocol/test/oud/oud-redeemer.ts

```diff
@@ -0,0 +1,154 @@
+        templeLineOfCredit = await new FakeTempleLineOfCredit__factory(owner).deploy();
+        oudRedeemer = await new OudRedeemer__factory(owner).deploy(
+            oudToken.address,
+            templeToken.address,
+            stableToken.address,
+            templeLineOfCredit.address,
+            TPI
+        );
+        await templeToken.addMinter(oudRedeemer.address);
+        await oudToken.addMinter(oudRedeemer.address);
+
+
+        return {
+            oudToken, templeToken, stableToken, oudRedeemer
+        };
+    }
+
+    beforeEach(async () => {
+        ({
+            oudToken, templeToken, stableToken, oudRedeemer
+        } = await loadFixture(setup));
+    });
+
+    describe("Admin", () => {
+        describe("Constructor Values", () => {
+            it("Returns correct TPI", async () => {
+                expect(await oudRedeemer.treasuryPriceIndex()).to.eq(TPI);
+                await oudRedeemer.connect(owner).setTreasuryPriceIndex(1400);
+                expect(await oudRedeemer.treasuryPriceIndex()).to.eq(1400);
+
+            });
+            it("Returns correct precision", async () => {
+                expect(await oudRedeemer.TPI_PRECISION()).to.eq(4);
+            });
+        });
+        it("Only owner can set TPI", async () => {
+            await expect(oudRedeemer.connect(alan).setTreasuryPriceIndex(2000))
+                .to.revertedWith("Ownable: caller is not the owner");
+            await oudRedeemer.connect(owner).setTreasuryPriceIndex(2000);
+            expect(await oudRedeemer.treasuryPriceIndex()).to.eq(2000);
+        });
+        it("Only owner can set Stable coin to valid address", async () => {
+            await expect(oudRedeemer.connect(alan).setStableCoin(templeToken.address))
+                .to.revertedWith("Ownable: caller is not the owner");
+            await expect(oudRedeemer.connect(owner).setStableCoin(ZERO_ADDRESS))
+                .to.be.revertedWithCustomError(oudRedeemer,"AddressZero");
+            await oudRedeemer.connect(owner).setStableCoin(templeToken.address);
+            expect(await oudRedeemer.stable()).to.eq(getAddress(templeToken.address));
+        });
+    });
+
+    describe("Redemption TXS", () => {
+        it("Returns the correct quote given Oud amount", async () => {
+            const [ _oudAmount, _oudTokenAddress, _stableAmount, _stableAddress, _templeAmount, _templeTokenAddress] 
+                = await oudRedeemer.getQuote(toAtto(OUD_AMOUNT));
+            expect(_oudAmount).to.eq(toAtto(OUD_AMOUNT));
+            expect(_oudTokenAddress).to.eq(oudToken.address);
+            expect(_stableAmount).to.eq(toAtto(STABLE_AMOUNT));
+            expect(_stableAddress).to.eq(stableToken.address);
+            expect(_templeAmount).to.eq(toAtto(OUD_AMOUNT));
+            expect(_templeTokenAddress).to.eq(templeToken.address);
+        });
+        it("Correct amount of Oud is burned for redemption", async () => {
```

Comment:
Having as 4 separate tests is fine - but could also consider using (hardhat becomes sloooowwww with lots of tests...)

`helpers.expectBalancesChangeBy()`

Within the lambda, you can check the event is emitted, and can then also check all token balances that you need to

(This is more of a PSA in case it's useful in the future)

Reply:
ack